services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This service is responsible for auto-detecting the admin operation protocol version for all clusters periodically.
+ * Each time ProtocolVersionDetectionTask runs, it will:
+ * 1. Get the current admin operation protocol versions from all controllers (parent + child) in the current cluster
+ * and find the smallest version - good version to use.
+ * 2. Get the admin operation protocol version from ZK.
+ * 3. If the version in ZK is different from the current version, update the version in ZK.
+ * To disable step 3, set the admin version for the cluster in ZK as -1.
+ */
+public class ProtocolVersionAutoDetectionService extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(ProtocolVersionAutoDetectionService.class);
+  private final AtomicBoolean stop = new AtomicBoolean(false);
+  private final Set<String> allClusters;
+  private final Map<String, ProtocolVersionAutoDetectionStats> clusterToStatsMap = new HashMap<>();
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceControllerMultiClusterConfig multiClusterConfigs;
+  final ScheduledExecutorService executor;
+
+  public ProtocolVersionAutoDetectionService(
+      VeniceParentHelixAdmin parentHelixAdmin,
+      VeniceControllerMultiClusterConfig multiClusterConfigs,
+      MetricsRepository metricsRepository,
+      Optional<Map<String, ProtocolVersionAutoDetectionStats>> clusterToStatsMap) {
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+    this.multiClusterConfigs = multiClusterConfigs;
+    this.allClusters = multiClusterConfigs.getClusters();
+    executor = Executors
+        .newScheduledThreadPool(multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionThreadCount());
+
+    if (clusterToStatsMap.isPresent()) {
+      // If the caller has provided a map of stats, use it - for testing purpose
+      this.clusterToStatsMap.putAll(clusterToStatsMap.get());
+    } else {
+      // Otherwise, create a new map of stats for each cluster - default option
+      for (String clusterName: this.allClusters) {
+        this.clusterToStatsMap.put(
+            clusterName,
+            new ProtocolVersionAutoDetectionStats(
+                metricsRepository,
+                "admin_operation_protocol_version_auto_detection_service_" + clusterName));
+      }
+    }
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    executor.scheduleAtFixedRate(
+        new ProtocolVersionDetectionTask(),
+        0,
```

Comment:
I would recommend giving it a 30 seconds initial delay. However, this might affect existing test cases; try to make a commit with this change only, and see if tests break :D

Reply:
Or, instead of using 30s, use the same value from multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionIntervalMs().

Reply:
I sleep the thread for getAdminOperationProtocolVersionAutoDetectionIntervalMs() and then run. Because I don't use executor.scheduleAtFixedRate() anymore, I don't exactly have the delay :D